### PR TITLE
[Core] Proposal to solve #4361

### LINKED ISSUE #4361: Math Utils are not compiling with AMatrix
@pooyan-dadvand @philbucher @loumalouomega 
I tried today to compile the core using AMatrix and I got an error in math utils, more specifically in the InvertMatrix method. This method uses a template in order to be able to use it with both BoundedMatrix and Matrix. Besides, it uses inside a resize to make sure that sizes of original and inverted matrices are the same. The problem is that AMatrix does not allowed to do a resize of a BoundedMatrix as it is considered an inefficiency.
Maybe cleanest solution would be to throw an error when sizes of input and output matrices do not match. What do you think?

### PR DESCRIPTION
**Pull request by loumalouomega**
Fixes #4361

MathUtils are not compiling with AMatrix, this solution proposes a transition solution

PS: #4366 Was not targeting the master to merge

### CODE REVIEW

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
I am blocking since this change would break the existing behaviour

**Comment by loumalouomega** [comment]
> I am blocking since this change would break the existing behaviour

I know this is not a fix. It was discussed on #4366

File: kratos/utilities/math_utils.h [hunk 1]
```diff
@@ -392,9 +394,9 @@ class MathUtils
         )
     {
 #ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it
-        AMatrix::LUFactorization<MatrixType, DenseVector<std::size_t> > lu_factorization(A);
+        AMatrix::LUFactorization<MatrixType, DenseVector<std::size_t> > lu_factorization(rA);
```
Comment (adityaghantasala) [inline kratos/utilities/math_utils.h]: This is just `A` not `rA`

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: Done

**Comment by adityaghantasala** [comment]
@loumalouomega  I see that it passes all the tests right ? Is it because that the macro `KRATOS_USE_AMATRIX` is not used in CD CI process ? If that is the case, there is no way to test it right  ?

**Comment by loumalouomega** [comment]
> 
> 
> @loumalouomega I see that it passes all the tests right ? Is it because that the macro `KRATOS_USE_AMATRIX` is not used in CD CI process ? If that is the case, there is no way to test it right ?

Yep

**Comment by GuillermoCasas** [comment]
The members of @KratosMultiphysics/implementation-committee have had a discussion about this issue and have not been able to reach a strong conclusion since there are arguments in favour (The PR compiles and it is a minor change) and against (There are currently no automatized compilations using the flag `KRATOS_USE_AMATRIX`, so the PR is not really passing any test for that). We thus request the @KratosMultiphysics/technical-committee for input on this one.

File: kratos/utilities/math_utils.h [hunk 2]
```diff
@@ -429,12 +432,16 @@ class MathUtils
         const TDataType Tolerance = ZeroTolerance
         )
     {
-        const SizeType size = rInputMatrix.size2();
+        const SizeType size = rInputMatrix.size1();
```
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h:435]: why is this changed from size2 to size1?

**Review by RiccardoRossi — COMMENTED** [review]
this has been in the queue for so long that we don't remember the details of this. 

we'll put it back to the todos

File: kratos/utilities/math_utils.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h:435]: I do not remember, too much time ago, but I know it was for a reason. Probably in order to consider not square matrices

**Comment by RiccardoRossi** [comment]
this looks like inactive. closing it, unless someone wants to revive it

**Comment by loumalouomega** [comment]
> this looks like inactive. closing it, unless someone wants to revive it

What is the status of Amatrix?
